### PR TITLE
refactor: move json parser into its own lib

### PR DESCRIPTION
**Pull request by olgavrou**
and move json tests to gtests

### CODE REVIEW

File: test/unit_test/CMakeLists.txt
Comment (jackgerrits) [inline test/unit_test/CMakeLists.txt]: These shouldnt be here

File: test/unit_test/CMakeLists.txt
Comment (olgavrou) [inline test/unit_test/CMakeLists.txt:27]: added during merge, will remove

File: vowpalwabbit/core/include/vw/core/parse_example_json.h
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/parse_example_json.h:14]: Needs `inline`

File: vowpalwabbit/core/src/parse_slates_example_json.cc
Comment (olgavrou) [inline vowpalwabbit/core/src/parse_slates_example_json.cc:4]: explicit template initialization needs to stick around here and in `parse_example_json.cc` until the functions are deprecated

File: vowpalwabbit/json_parser/include/vw/json_parser/json_utils.h
Comment (jackgerrits) [inline vowpalwabbit/json_parser/include/vw/json_parser/json_utils.h:11]: We should make json_utils a private header and move decision_service_interaction to a separate header that can be made public

File: vowpalwabbit/test_common/include/vw/test_common/test_common.h
Comment (olgavrou) [inline vowpalwabbit/test_common/include/vw/test_common/test_common.h:23]: parse_json/dsjson still used in most unit_tests so need to be available

File: vowpalwabbit/json_parser/CMakeLists.txt
Comment (jackgerrits) [inline vowpalwabbit/json_parser/CMakeLists.txt:13]: `vw_config` shouldnt be needed as a dep

File: vowpalwabbit/json_parser/tests/dsjson_parser_test.cc
Comment (jackgerrits) [inline vowpalwabbit/json_parser/tests/dsjson_parser_test.cc:122]: Replace with EXPECT_FLOAT_EQ
```suggestion
  EXPECT_FLOAT_EQ(0.0f, interaction.probability_of_drop);

```

File: vowpalwabbit/test_common/include/vw/test_common/test_common.h
Comment (jackgerrits) [inline vowpalwabbit/test_common/include/vw/test_common/test_common.h:16]: This shouldn't be needed in our test code anymore. Places which used to use it can become `EXPECT_FLOAT_EQ`

File: vowpalwabbit/json_parser/tests/dsjson_parser_test.cc
Comment (jackgerrits) [inline vowpalwabbit/json_parser/tests/dsjson_parser_test.cc:105]:
```suggestion
  EXPECT_FLOAT_EQ(0.1f, interaction.probability_of_drop);

```

File: vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h
Comment (jackgerrits) [inline vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h:33]: What does `_s` mean?

File: test/unit_test/CMakeLists.txt
Comment (olgavrou) [reply test/unit_test/CMakeLists.txt]: outdated

File: vowpalwabbit/core/include/vw/core/parse_example_json.h
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/parse_example_json.h:14]: templates are implicitly inline

File: vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h
Comment (olgavrou) [reply vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h:33]: I am guessing string, we could rename it to plain `read_line_json`

File: vowpalwabbit/core/include/vw/core/parse_example_json.h
Comment (jackgerrits) [reply vowpalwabbit/core/include/vw/core/parse_example_json.h:14]: Oh true!

File: vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h
Comment (jackgerrits) [reply vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h:33]: Maybe after this merged we can do a pass cleaning up the functions a bit? We can leave it to later to make this one easier though

Comment (olgavrou) [reply vowpalwabbit/json_parser/include/vw/json_parser/parse_example_json.h:33]: yes we should, I also wanted to make some char*/size -> string_view transistions but the PR is big already

File: vowpalwabbit/json_parser/include/vw/json_parser/json_utils.h
Comment (olgavrou) [reply vowpalwabbit/json_parser/include/vw/json_parser/json_utils.h]: moved json_utils into src and left the ds struct in decision_service_utils header

File: vowpalwabbit/json_parser/include/vw/json_parser/decision_service_utils.h
Comment (jackgerrits) [inline vowpalwabbit/json_parser/include/vw/json_parser/decision_service_utils.h:6]: Headers aren't needed

Comment (olgavrou) [reply vowpalwabbit/json_parser/include/vw/json_parser/decision_service_utils.h:6]: and missing copyright

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**
